### PR TITLE
Wrap grdsample

### LINKED ISSUE #1422: grdsample alternative in pygmt
gmt has grdsample function for sampling a grid file with inputs of grid spacing and other parameters.
But pygmt doesn't have grdsample. 

What can I use instead of grdsample in pygmt?

### PR DESCRIPTION
**Pull request by willschlitzer**
This pull requests wraps the GMT module [`grdsample`](https://docs.generic-mapping-tools.org/6.2/grdsample.html).

Fixes #1422

**Reminders**

- [ ] Run `make format` and `make check` to make sure the code follows the style guide.
- [ ] Add tests for new features or tests that would have caught the bug that you're fixing.
- [ ] Add new public functions/methods/classes to `doc/api/index.rst`.
- [ ] Write detailed docstrings for all functions/methods.
- [ ] If adding new functionality, add an example to docstrings or tutorials.

**Slash Commands**

You can write slash commands (`/command`) in the first line of a comment to perform
specific operations. Supported slash commands are:

- `/format`: automatically format and lint the code
- `/test-gmt-dev`: run full tests on the latest GMT development version

### CODE REVIEW

File: pygmt/src/grdsample.py
Comment (weiji14) [inline pygmt/src/grdsample.py:84]: I like this chunk of docstring. Should it go into https://github.com/GenericMappingTools/pygmt/blob/60a92b3ee2cc37cf9277a2fd623cffd2b52abcba/pygmt/helpers/decorators.py#L166-L175 perhaps so we can reuse it for other modules that support interpolation (-n) like `grdtrack` (as reported in https://github.com/GenericMappingTools/pygmt/issues/1378#issuecomment-875915899)?

File: pygmt/src/grdsample.py
Comment (weiji14) [inline pygmt/src/grdsample.py:87]: Typo.

```suggestion
        Set registration to **g**\ ridline or **p**\ ixel.
```

File: pygmt/src/grdsample.py
Comment (weiji14) [inline pygmt/src/grdsample.py:86]: Technically the registration (-r) option can also accept a boolean. With `registration=True` being equal to `registration='g'` (gridline).

File: pygmt/src/grdsample.py
Comment (weiji14) [inline pygmt/src/grdsample.py:89]: There are common aliases for `coltypes` (-f) and `cores` (-x).

```suggestion
    {V}
    {f}
    {x}
```

File: pygmt/src/grdsample.py
Comment (weiji14) [inline pygmt/src/grdsample.py:25]:
```suggestion
    f="coltypes",
    r="registration",
    n="interpolation",
    x="cores",
```

**Review by weiji14 — COMMENTED** [review]
Thanks @willschlitzer! Just a few suggestions regarding the use of common aliases.

File: pygmt/src/grdsample.py
Comment (willschlitzer) [reply pygmt/src/grdsample.py:86]: I haven't looked at the other implementations of `-n`; does it accept uniform arguments for all modules?

File: pygmt/src/grdsample.py
Comment (willschlitzer) [reply pygmt/src/grdsample.py]: added in 593d712

File: pygmt/src/grdsample.py
Comment (meghanrjones) [reply pygmt/src/grdsample.py:86]: The same format is used throughout the GMT documentation for all modules that accept the `-n` common option (i.e., [**b**|**c**|**l**|**n**][**+a**][**+b***BC*][**+c**][**+t***threshold*]).

Comment (willschlitzer) [reply pygmt/src/grdsample.py]: Updated in 2ff1550

File: pygmt/src/grdsample.py
Comment (weiji14) [inline pygmt/src/grdsample.py:67]: Just reordering `n` to be placed between `V` and `f`.

```suggestion
    registration : str or bool
        [**g**\ |\ **p**\ ].
        Set registration to **g**\ ridline or **p**\ ixel.
    {V}
    {f}
    {n}
```

File: pygmt/helpers/decorators.py
Comment (willschlitzer) [reply pygmt/helpers/decorators.py:192]: What happens when you try and finish your commit in a hurry! :man_facepalming:

**Review by weiji14 — APPROVED** [review]
Looks good to me. But please hold off on merging until we decide on whether we want a PyGMT v0.4.1 release, or go straight to v0.5.0 as discussed in https://github.com/GenericMappingTools/pygmt/issues/1356#issuecomment-875425806

**Comment by michaelgrund** [comment]
Do we still want to wait or merge before the 0.4.1 release ?

File: pygmt/helpers/decorators.py
Comment (michaelgrund) [inline pygmt/helpers/decorators.py:184]:
```suggestion
            - **+a** to switch off antialiasing (where supported) [Default uses
```
To be consistent with the stylte @meghanrjones used in #1407 😉.

File: pygmt/helpers/decorators.py
Comment (michaelgrund) [inline pygmt/helpers/decorators.py:191]:
```suggestion
              [Default may exceed limits].
```

File: pygmt/helpers/decorators.py
Comment (michaelgrund) [inline pygmt/helpers/decorators.py:196]:
```suggestion
              and 0.1 will go about 90% of the way [Default is 0.5].""",
```

**Comment by meghanrjones** [comment]
> Do we still want to wait or merge before the 0.4.1 release ?

I think we need to wait until after v0.4.1 in order to not break the versioning rules.

File: pygmt/src/grdsample.py
Comment (weiji14) [inline pygmt/src/grdsample.py:26]: n comes before r.
```suggestion
    n="interpolation",
    r="registration",
```

**Review by weiji14 — COMMENTED** [review]
Thanks for your patience @willschlitzer, now that the v0.4.1 tag is behind, we can finally finish this PR (and all the other ones you have lined up). Could you please resolve the conflict on `decorators.py` and fix one tiny issue I suggested below?

**Comment by willschlitzer** [comment]
@weiji14 I made the changes you suggested

File: pygmt/tests/test_grdsample.py
Comment (seisman) [inline pygmt/tests/test_grdsample.py:31]:
```suggestion
```

File: pygmt/tests/test_grdsample.py
Comment (seisman) [inline pygmt/tests/test_grdsample.py:27]: Using `increment=[1, 2.5]` gives this warning. I think it's better to use a smaller y increment, like `[1, 0.5]`.
```
grdsample [WARNING]: Output sampling interval in y exceeds input interval and may lead to aliasing.
```

**Review by seisman — APPROVED** [review]

File: pygmt/tests/test_grdsample.py
Comment (seisman) [reply pygmt/tests/test_grdsample.py:27]:
```suggestion
        result = grdsample(grid=grid, outgrid=tmpfile.name, increment=[1, 0.5])
```

File: pygmt/tests/test_grdsample.py
Comment (seisman) [inline pygmt/tests/test_grdsample.py:33]:
```suggestion
        assert float(result[7]) == 0.5  # y-increment
```

**Review by michaelgrund — APPROVED** [review]
Fine to merge after the suggestions by @seisman are implemented!

File: pygmt/src/grdsample.py
Comment (seisman) [inline pygmt/src/grdsample.py:20]: It seems other modules (e.g., `grdfilter`) use the alias `spacing` rather than `increment`. We should keep the alias consistent, no?

Comment (willschlitzer) [inline pygmt/src/grdsample.py:20]:
```suggestion
    I="spacing",
```

File: pygmt/src/grdsample.py
Comment (willschlitzer) [inline pygmt/src/grdsample.py:37]:
```suggestion
    ``spacing``, and set a new sub-region using ``region``. A bicubic
```

File: pygmt/src/grdsample.py
Comment (willschlitzer) [inline pygmt/src/grdsample.py:42]:
```suggestion
    the input grid. When ``spacing`` is omitted, the grid spacing of the
```

File: pygmt/tests/test_grdsample.py
Comment (willschlitzer) [inline pygmt/tests/test_grdsample.py:32]:
```suggestion
    grdsample with an outgrid set and the spacing is changed.
    """
    with GMTTempFile(suffix=".nc") as tmpfile:
        result = grdsample(grid=grid, outgrid=tmpfile.name, spacing=[1, 0.5])
        assert result is None  # return value is None
        assert os.path.exists(path=tmpfile.name)  # check that outgrid exists
        result = grdinfo(tmpfile.name, per_column=True).strip().split()
        assert float(result[6]) == 1  # x-increment
        assert float(result[7]) == 0.5  # y-increment
```

**Review by weiji14 — APPROVED** [review]
:rocket:

**Pull request merged**
